File: QuantConnect/Strategy_6.py


# Momentum_CCI.py

from QuantConnect.Data.Market import TradeBar
from QuantConnect.Indicators import CommodityChannelIndex
from QuantConnect.Algorithm.Framework import QCAlgorithmFramework
from QuantConnect.Algorithm.Framework.Execution import ExecutionModel
from QuantConnect.Algorithm.Framework.Portfolio import EqualWeightingPortfolioConstructionModel
from QuantConnect.Algorithm.Framework.Risk import MaximumDrawdownPercentPerSecurity
from datetime import timedelta


class Momentum(QCAlgorithmFramework):

    def Initialize(self):
        self.SetStartDate(2010, 1, 1)
        self.SetEndDate(2020, 1, 1)
        self.SetCash(100000)

        self.AddEquity("SPY", Resolution.Daily)

        self.cci = self.INDICATOR(CommodityChannelIndex, "SPY", 20)
        self.previous = None

        # Set up portfolio construction and risk management models
        self.SetPortfolioConstruction(EqualWeightingPortfolioConstructionModel())
        self.SetRiskManagement(MaximumDrawdownPercentPerSecurity(0.01))

    def OnData(self, data):
        
        if not (data.ContainsKey("SPY") and data["SPY"] and data["SPY"].Close):
            return
        
        current = data["SPY"].Close
        
        if not (self.previous and current):
            return
        
        holdings = sum([x.Value.Holdings.Quantity for x in self.Portfolio.Values])
        
        if holdings == 0:
            if self.cci.Current.Value > 100:
                # Buy signal above 100 level
                quantity = int(self.Portfolio.TotalPortfolioValue / current)
                order = self.Order("SPY", quantity)
                
                # Set stop loss at 1 x ATR below entry price
                atr = self.ATR("SPY", 20).Current.Value
                stopPrice = current - atr
                order.StopMarket(stopPrice)

            elif self.cci.Current.Value < -100:
                # Short signal below -100 level
                quantity = int(self.Portfolio.TotalPortfolioValue / current)
                order = self.Order("SPY", -quantity)
                
                # Set stop loss at 1 x ATR above entry price
                atr = self.ATR("SPY", 20).Current.Value
                stopPrice = current + atr
                order.StopMarket(stopPrice)

            else:
                return
                
            # Set previous price for next iteration    
            self.previous = current
            
        
        elif holdings > 0:
            
            if (self.cci.Current.Value < -100) or (current < order.StopPrice):
                
                # Sell signal or trailing stop loss triggered    
                sellQuantity = int(self.Portfolio.TotalPortfolioValue / current)
                
                if sellQuantity == holdings:
                    # Liquidate all shares if no partial sells allowed 
                    order.Liquidate()
                    
                    # Reset previous price for next iteration    
                    self.previous = None
                
                
        
        elif holdings < 0:
            
            if (self.cci.Current.Value > 100) or (current > order.StopPrice):
                
                 # Buy signal or trailing stop loss triggered    
                 buyQuantity= int(self.Portfolio.TotalPortfolioValue / current)
                 
                 if buyQuantity == abs(holdings):
                     # Liquidate all shares if no partial buys allowed 
                     order.Liquidate()
                     
                     # Reset previous price for next iteration    
                     self.previous = None
                
        
    def OnOrderEvent(self, orderEvent):
        
        if not(orderEvent.Status == OrderStatus.Filled or 
               orderEvent.Status == OrderStatus.PartiallyFilled): 
             return
        
         fillQuantity=orderEvent.FillQuantity
        
         fillPrice=orderEvent.FillPrice
        
         if fillQuantity>0: 
            
             buyOrder=self.Transactions.GetOrderById(orderEvent.OrderId) 
            
             stopLossPrice=fillPrice-self.ATR("SPY",20).Current.Value
            
             buyOrder.StopMarket(stopLossPrice) 
        
         elif fillQuantity<0: 
            
             sellOrder=self.Transactions.GetOrderById(orderEvent.OrderId) 
            
             stopLossPrice=fillPrice+self.ATR("SPY",20).Current.Value
            
             sellOrder.StopMarket(stopLossPrice)

File: QuantConnect/Strategy_6.cs


# Momentum.py
from QuantConnect.Data.Market import TradeBar
from QuantConnect.Indicators import CommodityChannelIndex, RollingWindow
from datetime import timedelta

class Momentum(QCAlgorithm):
    
    def Initialize(self):
        self.SetStartDate(2010, 1, 1)
        self.SetEndDate(2021, 8, 31)
        self.SetCash(100000)
        self.symbol = "SPY"
        self.AddEquity(self.symbol, Resolution.Daily)
        
        self.cci = CommodityChannelIndex(14)
        self.window = RollingWindow[TradeBar](2)
        
        self.stopMarketTicket = None
        self.entryPrice = None
        self.trailStopPrice = None
        
    def OnData(self, data):
        if not data.ContainsKey(self.symbol) or not data[self.symbol].HasData:
            return
        
        currentBar = data[self.symbol]
        
        # Update the indicators and rolling window with the current bar.
        self.cci.Update(currentBar.EndTime, currentBar.Close)
        
        if not (self.cci.IsReady and len(self.window) == 2):
            return
        
         # Check if we currently have a position
        if not self.Portfolio.Invested:
            if (self.cci.Current.Value > 100 and 
                self.window[0].Close > self.window[1].Close):
                
                # Buy the asset and set stop loss order.
                entryPrice = currentBar.Close
                stopMarketTicket = self.StopMarketOrder(
                    symbol=self.symbol,
                    quantity=self.CalculateOrderQuantity(self.symbol, 0.95),
                    stopPrice=entryPrice - (self.ATR(self.symbol, 14).Current.Value * 1),
                    tag="Stop Loss"
                )
                
                # Set trailing stop order with initial parameters.
                trailStopTicket = self.TrailingStopMarketOrder(
                    symbol=self.symbol,
                    quantity=self.Portfolio[self.symbol].Quantity,
                    trailingStopAmount=self.ATR(self.symbol, 14).Current.Value * 2,
                    tag="Trailing Stop"
                )
                
                # Store entry price and ticket for later reference.
                self.entryPrice = entryPrice
                self.stopMarketTicket = stopMarketTicket
                return
            
            elif (self.cci.Current.Value < -100 and 
                  self.window[0].Close < self.window[1].Close):
                  
                  # Short the asset and set stop loss order.
                  entryPrice = currentBar.Close
                  stopMarketTicket = self.StopMarketOrder(
                      symbol=self.symbol,
                      quantity=self.CalculateOrderQuantity(self.symbol, -0.95),
                      stopPrice=entryPrice + (self.ATR(self.Symbol, 14).Current.Value * 1),
                      tag="Stop Loss"
                  )
                  
                  # Set trailing stop order with initial parameters.
                  trailStopTicket = TrailingStopMarketOrder(
                      symbol=self.Symbol,
                      quantity=self.Portfolio[self.Symbol].Quantity * -1,
                      trailingStopAmount=self.ATR(Symbol, 14).Current.Value *2,
                      tag="Trailing Stop"
                  )
                  
                  # Store entry price and ticket for later reference.
                  entryPrice = currentBar.Close
                  stopMarketTicket = stopMarketTicket   
        
         else:
             # Check exit conditions.
             if ((self.Portfolio[self.Symbol].IsLong and 
                 ((self.cci.Current.Value < -100 and 
                   (currentBar.Close < max([x.Close for x in window])))) or 
                 (self.Portfolio[self.Symbol].IsShort and 
                 ((self.cci.Current.Value >100)and(currentbar.close>min([x.close for x in window]))))):
                 
                 # Close position and cancel pending orders.
                 cancelOrdersResult=CancelOpenOrders()
                 
                 if cancelOrdersResult.IsSuccess:
                     selling_price=currentbar.close
                     position_qty=-self.Portfolio[self.Symbol].Quantity \
                         if Portfolio[self.Symbol] .IsShort else Portfolio [symbol] .Quantity                     
                     limit_price=selling_price \
                         +position_qty*(entry_price-self.stop_loss.Price)/abs(position_qty) \
                         if Portfolio[symbol] .IsLong else selling_price \
                         +position_qty*(entry_price+stop_loss.Price)/abs(position_qty)                     
                     
                     limit_price=max(limit_price,currentbar.Low) \
                         if Portfolio[symbol] .IsLong else min(limit_price,currentbar.High)
                     
                     ticket=LimitOrder(symbol,self.CalculateOrderQuantity(symbol,-0.5*position_qty),limit_price,"Exit")\
                        if Portfolio[symbol] .IsLong else LimitOrder(symbol,self.CalculateOrderQuantity(symbol,-position_qty),limit_price,"Exit")
                        
                        
                    
                
                
                 
                
                
                
             
             
             
             
             
             
             
             
             
        
        
                
                
        
        
        
        
        
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    



        

    
        
        
        
        
        
        
        
        
